upf/dnf.py

```diff
@@ -0,0 +1,179 @@
+                    # iteration.
+                    stack.append((p, new_e, True))
+                    stack.append((p, e1, True))
+                    stack.append((p, e.arg(0), False))
+                    stack.append((p, e.arg(1), False))
+                    stack.append((p, e2, True))
+                    stack.append((not p, e.arg(0), False))
+                    stack.append((not p, e.arg(1), False))
+                else:
+                    if p:
+                        solved.append(e)
+                    else:
+                        solved.append(self.manager.Not(e))
+        return solved.pop()
+
+
+
+
+class Dnf(DagWalker):
+    """Class used to transform a logic expression into the equivalent
+    Disjunctive Normal Form expression.
+
+    This is done first by transforming the expression into a NNF expression,
+    and then every And and Or are propagated to be a unique equivalent Or of
+    Ands or Atomic expressions, where 'atomic expressions' could also be a
+    Not of an atomic expression.
+    """
+    def __init__(self, env: 'upf.environment.Environment'):
+        DagWalker.__init__(self, True)
+        self.env = env
+        self.manager = env.expression_manager
+        self._nnf = Nnf(self.env)
+
+    def get_dnf_expression(self, expression: FNode) -> FNode:
+        """Function used to transform a logic expression into the equivalent
+        Disjunctive Normal Form expression.
+
+        This is done first by transforming the expression into a NNF expression,
+        and then every And and Or are propagated to be a unique equivalent Or of
+        Ands or Atomic expressions, where 'atomic expressions' could also be a
+        Not of an atomic expression.
+
+        For example, the form: !(a => (b && c)) becomes:
+        a && (!b || !c), in NNF form, and then:
+        (a && !b) || (a && !c), therefore a DNF expression."""
+        nnf_exp = self._nnf.get_nnf_expression(expression)
+        tuples = self.walk(nnf_exp)
+        and_list: List[FNode] = []
+        for and_args in tuples:
+            and_list.append(self.manager.And(and_args))
+        return self.manager.Or(and_list)
+
+    def walk_and(self, expression: FNode, args: List[List[List[FNode]]], **kwargs) -> List[List[FNode]]:
+        res: List[List[FNode]] = []
+        tuples = product(*args)
+        for na in tuples:
+            nl: List[FNode] = []
+            to_add = True
+            for nat in na:
+                for ne in nat:
+                    if ne.is_not():
+                        if ne.arg(0) in nl:
+                            to_add = False
+                            break
+                    else:
+                        if self.manager.Not(ne) in nl:
+                            to_add = False
+                            break
+                    if ne not in nl:
+                        nl.append(ne)
+                if not to_add:
+                    break
+            if to_add:
```

Comment:
Too much cognitive complexity for 6pm :smiley: 
@Framba-Luca What is this piece of code trying to achieve?

General Feedback: this code is hard to read because of the nesting of ifs/fors and because variable names are not too descriptive.

Reply:
@mikand Ok, i tried to improve variables names and to add some comments with one example, let me know if it is clearer now.